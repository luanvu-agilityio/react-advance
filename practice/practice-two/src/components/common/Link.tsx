import type { ReactNode, MouseEvent, CSSProperties } from 'react'
import styled from 'styled-components'

interface LinkProps {
  href: string
  children: ReactNode
  onClick: (event: MouseEvent<HTMLAnchorElement>) => void
  className?: string
  target?: '_blank' | '_self' | '_parent' | '_top'
  disabled?: boolean
  style?: CSSProperties
}

const StyledLink = styled.a<{ disabled?: boolean }>`
  text-decoration: none;
  color: var(--green-color-default);
  cursor: ${({ disabled }) => (disabled ? 'not-allowed' : 'pointer')};
  opacity: ${({ disabled }) => (disabled ? 0.5 : 1)};

  &:hover {
    text-decoration: ${({ disabled }) => (disabled ? 'none' : 'underline')};
  }
`

const Link = ({
  href = '#',
  children,
  onClick,
  className,
  target = '_self',
  disabled = false,
  style,
}: LinkProps) => {
  const handleClick = (event: MouseEvent<HTMLAnchorElement>) => {
    if (disabled) {
      event.preventDefault()
      return
    }

    onClick(event)
  }

  return (
    <StyledLink
      href={href}
      onClick={handleClick}
      className={className}
      target={target}
      disabled={disabled}
      rel={target === '_blank' ? 'noopener noreferrer' : undefined}
      style={style}
    >
      {children}
    </StyledLink>
  )
}

export default Link
