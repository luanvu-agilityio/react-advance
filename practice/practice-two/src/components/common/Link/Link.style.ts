import styled from 'styled-components'

export const StyledLink = styled.a<{ disabled?: boolean }>`
  text-decoration: none;
  color: var(--green-color-default);
  cursor: ${({ disabled }) => (disabled ? 'not-allowed' : 'pointer')};
  opacity: ${({ disabled }) => (disabled ? 0.5 : 1)};

  &:hover {
    text-decoration: ${({ disabled }) => (disabled ? 'none' : 'underline')};
  }
`
