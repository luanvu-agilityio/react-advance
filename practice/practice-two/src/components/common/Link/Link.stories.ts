import type { Meta, StoryObj } from '@storybook/react'
import Link from './index'

const meta = {
  title: 'Components/Common/Link',
  component: Link,
  parameters: {
    layout: 'centered',
  },
  tags: ['autodocs'],
  argTypes: {
    href: { control: 'text' },
    children: { control: 'text' },
    onClick: { action: 'clicked' },
    className: { control: 'text' },

    disabled: { control: 'boolean' },
    style: { control: 'object' },
  },
} satisfies Meta<typeof Link>

export default meta
type Story = StoryObj<typeof Link>

export const Default: Story = {
  args: {
    href: 'https://google.com.vn',
    children: 'Google VN',
    onClick: (e: React.MouseEvent<HTMLAnchorElement>) => e.preventDefault(),
  },
}

export const ExternalLink: Story = {
  args: {
    href: 'https://google.com.vn',
    children: 'Google VN',

    onClick: (e: React.MouseEvent<HTMLAnchorElement>) => e.preventDefault(),
  },
}

export const DisabledLink: Story = {
  args: {
    href: 'https://google.com.vn',
    children: 'Google VN',
    disabled: true,
    onClick: (e: React.MouseEvent<HTMLAnchorElement>) => e.preventDefault(),
  },
}

export const CustomStyledLink: Story = {
  args: {
    href: 'https://google.com.vn',
    children: 'Google VN',
    onClick: (e: React.MouseEvent<HTMLAnchorElement>) => e.preventDefault(),
    style: {
      color: 'var(--green-color-default)',
      fontSize: '20px',
      fontWeight: 'bold',
    },
  },
}
