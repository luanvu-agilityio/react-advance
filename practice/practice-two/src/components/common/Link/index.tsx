import type { ReactNode, MouseEvent, CSSProperties } from 'react'
import { StyledLink } from './Link.style'

export interface LinkProps {
  href: string
  children: ReactNode
  onClick?: (event: MouseEvent<HTMLAnchorElement>) => void
  className?: string
  target?: '_blank' | '_self' | '_parent' | '_top'
  disabled?: boolean
  style?: CSSProperties
}

const Link = ({
  href = '#',
  children,
  onClick,
  className,
  target = '_self',
  disabled = false,
  style,
}: LinkProps) => {
  const handleClick = (event: MouseEvent<HTMLAnchorElement>) => {
    if (disabled) {
      event.preventDefault()
      return
    }

    onClick?.(event)
  }

  return (
    <StyledLink
      href={href}
      onClick={handleClick}
      className={className}
      target={target}
      disabled={disabled}
      rel={target === '_blank' ? 'noopener noreferrer' : undefined}
      style={style}
    >
      {children}
    </StyledLink>
  )
}

export default Link
