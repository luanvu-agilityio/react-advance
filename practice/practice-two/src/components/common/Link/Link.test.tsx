import { render, screen, fireEvent } from '@testing-library/react'
import Link from './index'

describe('Link Component', () => {
  const defaultProps = {
    href: 'https://example.com',
    children: 'Click me',
    onClick: jest.fn(),
  }

  beforeEach(() => {
    jest.clearAllMocks()
  })

<<<<<<< HEAD
  // SNAPSHOT TESTS FOR RENDERING
  describe('Rendering', () => {
    it('matches snapshots for various configurations', () => {
      // Default link
      const { container: defaultContainer } = render(<Link {...defaultProps} />)
      expect(defaultContainer).toMatchSnapshot('default-link')

      // Disabled link
      const { container: disabledContainer } = render(
        <Link {...defaultProps} disabled />
      )
      expect(disabledContainer).toMatchSnapshot('disabled-link')

      // Custom styled link
      const customStyle = { color: 'red', fontSize: '16px' }
      const { container: styledContainer } = render(
        <Link {...defaultProps} style={customStyle} />
      )
      expect(styledContainer).toMatchSnapshot('custom-styled-link')

      // External link (opens in new tab)
      const { container: externalContainer } = render(
        <Link href="https://example.com" target="_blank">
          External Link
        </Link>
      )
      expect(externalContainer).toMatchSnapshot('external-link')

      // Link with className
      const { container: classContainer } = render(
        <Link {...defaultProps} className="custom-class">
          Classy Link
        </Link>
      )
      expect(classContainer).toMatchSnapshot('link-with-class')
    })
  })

  // INTERACTIVE BEHAVIOR TESTS
  describe('Interactive Behavior', () => {
    it('handles click events when not disabled', () => {
      render(<Link {...defaultProps} />)
      const link = screen.getByText('Click me')

      fireEvent.click(link)
      expect(defaultProps.onClick).toHaveBeenCalledTimes(1)
    })

    it('prevents click events when disabled', () => {
      render(<Link {...defaultProps} disabled />)
      const link = screen.getByText('Click me')

      fireEvent.click(link)
      expect(defaultProps.onClick).not.toHaveBeenCalled()

      const preventDefaultSpy = jest.spyOn(Event.prototype, 'preventDefault')
      fireEvent.click(link)
      expect(preventDefaultSpy).toHaveBeenCalled()

      preventDefaultSpy.mockRestore()
=======
  test('renders correctly with default props', () => {
    render(<Link {...defaultProps} />)
    const link = screen.getByText('Click me')

    expect(link).toBeInTheDocument()
    expect(link).toHaveAttribute('href', 'https://example.com')
    expect(link).toHaveAttribute('target', '_self')
    expect(link).not.toHaveAttribute('rel')
  })

  test('handles click events when not disabled', () => {
    render(<Link {...defaultProps} />)
    const link = screen.getByText('Click me')

    fireEvent.click(link)
    expect(defaultProps.onClick).toHaveBeenCalled()
  })

  test('prevents click events when disabled', () => {
    render(<Link {...defaultProps} disabled />)
    const link = screen.getByText('Click me')

    fireEvent.click(link)
    expect(defaultProps.onClick).not.toHaveBeenCalled()
  })

  test('applies custom styles', () => {
    const customStyle = { color: 'red', fontSize: '16px' }
    render(<Link {...defaultProps} style={customStyle} />)
    const link = screen.getByText('Click me')

    expect(link).toHaveStyle(customStyle)
  })

  test('applies disabled styles', () => {
    render(<Link {...defaultProps} disabled />)
    const link = screen.getByText('Click me')

    expect(link).toHaveStyle({
      cursor: 'not-allowed',
      opacity: '0.5',
    })
  })

  test('handles different target values', () => {
    const targets = ['_self', '_blank', '_parent', '_top'] as const

    targets.forEach((target) => {
      const { rerender } = render(<Link {...defaultProps} target={target} />)
      const link = screen.getByText('Click me')

      expect(link).toHaveAttribute('target', target)
      rerender(<Link {...defaultProps} />)
>>>>>>> 7a35d9791a5da6fe80ff0a8541efaf78233de04d
    })
  })
})
