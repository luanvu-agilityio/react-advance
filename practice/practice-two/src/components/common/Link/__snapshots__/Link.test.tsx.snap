// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Link Component Rendering matches snapshots for various configurations: custom-styled-link 1`] = `
<div>
  <a
    class="sc-Qotzb jUyYcI"
    href="https://example.com"
    style="color: red; font-size: 16px;"
    target="_self"
  >
    Click me
  </a>
</div>
`;

exports[`Link Component Rendering matches snapshots for various configurations: default-link 1`] = `
<div>
  <a
    class="sc-Qotzb jUyYcI"
    href="https://example.com"
    target="_self"
  >
    Click me
  </a>
</div>
`;

exports[`Link Component Rendering matches snapshots for various configurations: disabled-link 1`] = `
<div>
  <a
    class="sc-Qotzb kzgBsp"
    disabled=""
    href="https://example.com"
    target="_self"
  >
    Click me
  </a>
</div>
`;

exports[`Link Component Rendering matches snapshots for various configurations: external-link 1`] = `
<div>
  <a
    class="sc-Qotzb jUyYcI"
    href="https://example.com"
    rel="noopener noreferrer"
    target="_blank"
  >
    External Link
  </a>
</div>
`;

exports[`Link Component Rendering matches snapshots for various configurations: link-with-class 1`] = `
<div>
  <a
    class="sc-Qotzb jUyYcI custom-class"
    href="https://example.com"
    target="_self"
  >
    Classy Link
  </a>
</div>
`;
